spaceinvaders.py

```diff
@@ -1,648 +1,687 @@
+from os.path import abspath, dirname
+from random import randrange, choice
+
+BASE_PATH = abspath(dirname(__file__))
+FONT_PATH = BASE_PATH + "/fonts/"
+IMAGE_PATH = BASE_PATH + "/images/"
+SOUND_PATH = BASE_PATH + "/sounds/"
 
 #           R    G    B
-WHITE 	= (255, 255, 255)
-GREEN 	= (78, 255, 87)
-YELLOW 	= (241, 255, 0)
-BLUE 	= (80, 255, 239)
-PURPLE 	= (203, 0, 255)
-RED 	= (237, 28, 36)
-
-SCREEN 		= display.set_mode((800,600))
-FONT = "fonts/space_invaders.ttf"
-IMG_NAMES 	= ["ship", "ship", "mystery", "enemy1_1", "enemy1_2", "enemy2_1", "enemy2_2",
-				"enemy3_1", "enemy3_2", "explosionblue", "explosiongreen", "explosionpurple", "laser", "enemylaser"]
-IMAGES 		= {name: image.load("images/{}.png".format(name)).convert_alpha()
-				for name in IMG_NAMES}
+WHITE = (255, 255, 255)
+GREEN = (78, 255, 87)
+YELLOW = (241, 255, 0)
+BLUE = (80, 255, 239)
+PURPLE = (203, 0, 255)
+RED = (237, 28, 36)
+
+SCREEN = display.set_mode((800, 600))
+FONT = FONT_PATH + "space_invaders.ttf"
+IMG_NAMES = ["ship", "ship", "mystery",
+             "enemy1_1", "enemy1_2",
+             "enemy2_1", "enemy2_2",
+             "enemy3_1", "enemy3_2",
+             "explosionblue", "explosiongreen", "explosionpurple",
+             "laser", "enemylaser"]
+IMAGES = {name: image.load(IMAGE_PATH + "{}.png".format(name)).convert_alpha()
+          for name in IMG_NAMES}
+
 
 class Ship(sprite.Sprite):
-	def __init__(self):
-		sprite.Sprite.__init__(self)
-		self.image = IMAGES["ship"]
-		self.rect = self.image.get_rect(topleft=(375, 540))
-		self.speed = 5
+    def __init__(self):
+        sprite.Sprite.__init__(self)
+        self.image = IMAGES["ship"]
+        self.rect = self.image.get_rect(topleft=(375, 540))
+        self.speed = 5
 
-	def update(self, keys, *args):
-		if keys[K_LEFT] and self.rect.x > 10:
-			self.rect.x -= self.speed
-		if keys[K_RIGHT] and self.rect.x < 740:
-			self.rect.x += self.speed
-		game.screen.blit(self.image, self.rect)
+    def update(self, keys, *args):
+        if keys[K_LEFT] and self.rect.x > 10:
+            self.rect.x -= self.speed
+        if keys[K_RIGHT] and self.rect.x < 740:
+            self.rect.x += self.speed
+        game.screen.blit(self.image, self.rect)
 
 
 class Bullet(sprite.Sprite):
-	def __init__(self, xpos, ypos, direction, speed, filename, side):
-		sprite.Sprite.__init__(self)
-		self.image = IMAGES[filename]
-		self.rect = self.image.get_rect(topleft=(xpos, ypos))
-		self.speed = speed
-		self.direction = direction
-		self.side = side
-		self.filename = filename
-
-	def update(self, keys, *args):
-		game.screen.blit(self.image, self.rect)
-		self.rect.y += self.speed * self.direction
-		if self.rect.y < 15 or self.rect.y > 600:
-			self.kill()
+    def __init__(self, xpos, ypos, direction, speed, filename, side):
+        sprite.Sprite.__init__(self)
+        self.image = IMAGES[filename]
+        self.rect = self.image.get_rect(topleft=(xpos, ypos))
+        self.speed = speed
```

Comment:
It's been a while since I've looked at this code... but I wonder if you can set the initial value of these variables in the constructor instead of having these redundant lines? Doesn't need to be done here but just a thought.

Reply:
I don't understand. Did you mean **kwargs and setattr or something else?
https://stackoverflow.com/questions/739625/setattr-with-kwargs-pythonic-or-not